### PR TITLE
JUnit codegen

### LINKED ISSUE #1039: [Feature] add junit to playwright codegen
similar to https://github.com/microsoft/playwright-dotnet/issues/2247

### PR DESCRIPTION
**Pull request by jfgreffier**
Codegen for JUnit

Following JUnit5 Playwright fixtures by @uchagani https://github.com/microsoft/playwright-java/issues/1369

Fixes https://github.com/microsoft/playwright-java/issues/1039

### CODE REVIEW

File: packages/playwright-core/src/server/recorder/java.ts [hunk 1]
```diff
@@ -26,19 +26,37 @@ import { JavaScriptFormatter } from './javascript';
...
     const action = actionInContext.action;
     const pageAlias = actionInContext.frame.pageAlias;
-    const formatter = new JavaScriptFormatter(6);
+    const offset = this._mode === 'junit' ? 4 : 6;
+    const formatter = new JavaScriptFormatter(offset);
 
     if (action.name === 'openPage') {
-      formatter.add(`Page ${pageAlias} = context.newPage();`);
+      if (this._mode !== 'junit' || pageAlias !== 'page')
+        formatter.add(`Page ${pageAlias} = context.newPage();`);
```
Comment (yury-s) [inline packages/playwright-core/src/server/recorder/java.ts:59]: In this case we also need to add `BrowserContext context` parameter to the test function.

**Review by yury-s — COMMENTED** [review]
The PR looks good modulo the comment. We are going to cut branch for 1.39 release soon and I'm not sure if JUnit fixtures will be ready for that release (another API review,  docs and tests are still to be done), so I'd hold this off landing this PR for now.

**Comment by jfgreffier** [comment]
> The PR looks good modulo the comment. We are going to cut branch for 1.39 release soon and I'm not sure if JUnit fixtures will be ready for that release (another API review, **docs and tests are still to be done**), so I'd hold this off landing this PR for now.

Understand. Docs and tests will probably need some time, indeed.

**Comment by uchagani** [comment]
@yury-s what docs/tests do we need? I'm happy to help with it.

File: packages/playwright-core/src/server/recorder/java.ts [hunk 2]
```diff
@@ -26,16 +26,36 @@ import { JavaScriptFormatter } from './javascript';
...
+  }
 
   generateAction(actionInContext: ActionInContext): string {
     const action = actionInContext.action;
     const pageAlias = actionInContext.frame.pageAlias;
-    const formatter = new JavaScriptFormatter(6);
+    const offset = this._mode === 'junit' ? 4 : 6;
+    const formatter = new JavaScriptFormatter(offset);
+
+    if (this._mode !== 'library' && (action.name === 'openPage' || action.name === 'closePage'))
```
Comment (jfgreffier) [inline packages/playwright-core/src/server/recorder/java.ts:57]: Similar to CSharp and other languages

File: packages/playwright-core/src/server/recorder/java.ts (hunk 1 above)
Comment (jfgreffier) [reply packages/playwright-core/src/server/recorder/java.ts]: After looking at other generators, I did the same thing in the case it is a test : nothing.  
Means `openpage` and and `closepage` are not managed, in the same way as tests in other languages.

We can improve this in another issue/PR. WDYT ?

**Comment by yury-s** [comment]
> @yury-s what docs/tests do we need? I'm happy to help with it.

By tests I meant switching our test harness in java to the new fixtures. I didn't want you to include it in the same PR with the implementation but I feel like it'd be nice to have as a validation of the approach.

By the docs I mean something similar to [this page](https://playwright.dev/dotnet/docs/test-runners#customizing-browserlaunch-options) and [this page](https://playwright.dev/python/docs/test-runners#examples).

**Comment by yury-s** [comment]
> By tests I meant switching our test harness in java to the new fixtures. I didn't want you to include it in the same PR with the implementation but I feel like it'd be nice to have as a validation of the approach.

To avoid possible confusion. Considering the idea of using a [Config fixture](https://github.com/microsoft/playwright-java/issues/1369#issuecomment-1745884063), it may be worth waiting until we commit to one approach or the other before porting the tests to avoid double work.

**Comment by pavelfeldman** [comment]
Closing as stale

**Pull request closed without merging**